components/webgpu/lib.rs

```diff
@@ -296,20 +314,69 @@ impl WGPU {
             receiver,
             sender,
             script_sender,
-            global: wgpu::hub::Global::new("wgpu-core", factory),
+            global: wgpu::hub::Global::new("wgpu-core", factory, wgt::BackendBit::PRIMARY),
             adapters: Vec::new(),
             devices: Vec::new(),
             _invalid_adapters: Vec::new(),
+            buffer_maps: HashMap::new(),
             webrender_api: webrender_api_sender.create_api(),
             webrender_document,
             external_images,
             wgpu_image_map,
         }
     }
 
-    fn run(mut self) {
+    fn run(&'a mut self) {
         while let Ok(msg) = self.receiver.recv() {
+            self.devices.iter().for_each(|device| {
+                let global = &self.global;
+                let device_id = device.0;
+                gfx_select!(device_id => global.device_poll(device_id, false));
+            });
             match msg {
+                WebGPURequest::BufferMapAsync {
+                    sender,
+                    buffer_id,
+                    host_map,
+                    map_range,
+                } => {
+                    let map_info = BufferMapInfo {
+                        buffer_id,
+                        sender,
+                        global: &self.global,
+                        size: (map_range.end - map_range.start) as usize,
+                    };
+                    self.buffer_maps.insert(buffer_id, map_info);
+
+                    unsafe extern "C" fn callback(status: BufferMapAsyncStatus, userdata: *mut u8) {
+                        let info = (userdata as *mut BufferMapInfo).as_ref().unwrap();
```

Comment:
This looks totally unsafe, the lifetime of `a` can't be respected if the stuff is unsafely casted from C.
Maybe just turn it into a `NonNull` pointer?